### DISCUSSION TITLE
Is it working with React Native?

### DISCUSSION THREAD

**Discussion by sharpordie** [yume-chan/ya-webadb#429] [Q&A]
Can we use this library into a React Native application to control another Android device?

**Comment by yume-chan** [comment]
This library only implements the ADB protocol and requires different backends on each platform to communicate with devices.

If there is a library that provides general USB access for React Native, it's easy to adapt that. Otherwise, it won't work.

**Reply by sharpordie** [reply]
I will rephrase my request, sorry if I was kinda evasive. I would like to send keyevents from my phone to my nvidia shield without adb binary.

This is what the [dadb library](https://github.com/mobile-dev-inc/dadb) do, but it's written in kotlin. Can ya-webadb do the same and be used in react-native application?

**Reply by yume-chan** [reply]
dadb is an ADB protocol implementation plus a TCP connector written in Kotlin. This library is only the ADB protocol implementation, it only knows how to parse and create ADB packets, not how to connect to your devices.

I also made connectors using WebUSB (for USB) and DirectSocket (for TCP) API in separated packages, but they only work in browsers, not React Native.

If you can make a connector for React Native (for example, a TCP one with packages like [react-native-tcp-socket](https://www.npmjs.com/package/react-native-tcp-socket)), then it will work.

**Reply by sharpordie** [reply]
Glad to know it could be doable. Unfortunately I don't have the knowledge. Even JavaScript itself is cryptic for me sometimes. I guess you are not interested in creating this backend yourself?

**Reply by yume-chan** [reply]
I also never used React Native. Creating a new backend (connector) is quick and easy, but I don't want to install a whole Android/iOS toolchain for React Native to test that. If you are not familiar with JavaScript and its ecosystem, creating a React Native project and integrating this package into that may also be a problem.

**Reply by lanzorg** [reply]
Hi yume-chan,
I am very interested about a react-native solution. When you say it would be easy and quick to create a react-native backend/connector, it can't be just cloning @yume-chan/adb-backend-ws package and replacing WebSocket with TcpSocket from react-native-tcp-socket? If that's the case I'll try it right away.

**Reply by yume-chan** [reply]
Yes, that's what you need. Be sure to check the `v0.0.16` tag (https://github.com/yume-chan/ya-webadb/blob/v0.0.16/libraries/adb-backend-ws/src/index.ts) if you want to use released NPM packages. (`main` branch separated stream related things to its own package.)

Just note that these packages use many newer JavaScript language features. As I said before I never used React Native, but as long as the JavaScript runtime is new enough, or if you can transpile them down, there shouldn't be any problem.
